File: mdix-java/src/main/java/com/midmanstudio/dixscript/ValueType.java
// ValueType.java
package com.midmanstudio.dixscript;

/**
 * DixScript value type discriminants, returned by {@link Database#valueTypeAt(String)}.
 * Ordinal values match the integers returned by the native getType() function.
 */
public enum ValueType {
    UNKNOWN(-1),
    NULL(0),
    BOOL(1),
    INT(2),
    LONG(3),
    FLOAT(4),
    DOUBLE(5),
    STRING(6),
    DATE(7),
    TIMESTAMP(8),
    HEX_COLOR(9),
    BLOB(10),
    REGEX(11),
    ARRAY(12),
    OBJECT(13),
    TUPLE(14),
    ENUM(15);

    private final int code;

    ValueType(int code) { this.code = code; }

    public int getCode() { return code; }

    public static ValueType fromCode(int code) {
        for (ValueType t : values()) {
            if (t.code == code) return t;
        }
        return UNKNOWN;
    }
}

File: mdix-php/src/Mdix/ValueType.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix;

/**
 * DixScript value type discriminants.
 *
 * Ordinal values match the integers returned by mdix_get_type() in the
 * native library (and MDIX_TYPE_* constants in the C header).
 *
 * FIX: this enum was previously missing the Long case entirely, which
 * shifted every case from Float onward one below its real native
 * discriminant (native Long=3 read back as this enum's old Float=3;
 * native Enum=15 had no matching case at all, since this enum only went
 * up to 14 — every actual DixScript enum field crashed valueTypeAt() with
 * an uncaught \ValueError). Values below are verified against
 * mdix-ffi/src/lib.rs's MdixType exactly.
 */
enum ValueType: int
{
    case Unknown   = -1;
    case Null      =  0;
    case Bool      =  1;
    case Int       =  2;
    case Long      =  3;
    case Float     =  4;
    case Double    =  5;
    case String    =  6;
    case Date      =  7;
    case Timestamp =  8;
    case HexColor  =  9;
    case Blob      = 10;
    case Regex     = 11;
    case Array     = 12;
    case Object    = 13;
    case Tuple     = 14;
    case Enum      = 15;

    /**
     * Returns a human-readable label, e.g. "int", "long", "string", "array".
     */
    public function label(): string
    {
        return match ($this) {
            self::Unknown   => 'unknown',
            self::Null      => 'null',
            self::Bool      => 'bool',
            self::Int       => 'int',
            self::Long      => 'long',
            self::Float     => 'float',
            self::Double    => 'double',
            self::String    => 'string',
            self::Date      => 'date',
            self::Timestamp => 'timestamp',
            self::HexColor  => 'hex_color',
            self::Blob      => 'blob',
            self::Regex     => 'regex',
            self::Array     => 'array',
            self::Object    => 'object',
            self::Tuple     => 'tuple',
            self::Enum      => 'enum',
        };
    }

    public function isScalar(): bool
    {
        return match ($this) {
            self::Bool, self::Int, self::Long, self::Float,
            self::Double, self::String => true,
            default => false,
        };
    }
}
